larpandoradlcontent/LArThreeDReco/LArEventBuilding/MLPBaseHierarchyTool.cc

```diff
@@ -0,0 +1,187 @@
+/**
+ *  @file   larpandoradlcontent/LArThreeDReco/LArEventBuilding/MLPBaseHierarchyTool.cc
+ *
+ *  @brief  Implementation of the MLP base hierarchy tool
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "Pandora/StatusCodes.h"
+
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+
+#include "larpandoradlcontent/LArHelpers/LArDLHelper.h"
+#include "larpandoradlcontent/LArThreeDReco/LArEventBuilding/MLPBaseHierarchyTool.h"
+
+#include <torch/script.h>
+#include <torch/torch.h>
+
+using namespace pandora;
+using namespace lar_content;
+
+namespace lar_dl_content
+{
+
+MLPBaseHierarchyTool::MLPBaseHierarchyTool() :
+    m_bogusFloat(-999.f),
+    m_vertexRegionRadius(5.f),
+    m_pfoListNames({"TrackParticles3D", "ShowerParticles3D"}),
+    m_detectorMinX(std::numeric_limits<float>::max()),
+    m_detectorMaxX(-std::numeric_limits<float>::max()),
+    m_detectorMinY(std::numeric_limits<float>::max()),
+    m_detectorMaxY(-std::numeric_limits<float>::max()),
+    m_detectorMinZ(std::numeric_limits<float>::max()),
+    m_detectorMaxZ(-std::numeric_limits<float>::max()),
+    areBoundariesSet(false)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void MLPBaseHierarchyTool::SetDetectorBoundaries()
+{
+    if (areBoundariesSet)
+        return;
+
+    const LArTPCMap &larTPCMap(this->GetPandora().GetGeometry()->GetLArTPCMap());
+
+    for (const auto &entry : larTPCMap)
+    {
+        m_detectorMinX = std::min(m_detectorMinX, entry.second->GetCenterX() - (entry.second->GetWidthX() * 0.5f));
+        m_detectorMaxX = std::max(m_detectorMaxX, entry.second->GetCenterX() + (entry.second->GetWidthX() * 0.5f));
+        m_detectorMinY = std::min(m_detectorMinY, entry.second->GetCenterY() - (entry.second->GetWidthY() * 0.5f));
+        m_detectorMaxY = std::max(m_detectorMaxY, entry.second->GetCenterY() + (entry.second->GetWidthY() * 0.5f));
+        m_detectorMinZ = std::min(m_detectorMinZ, entry.second->GetCenterZ() - (entry.second->GetWidthZ() * 0.5f));
+        m_detectorMaxZ = std::max(m_detectorMaxZ, entry.second->GetCenterZ() + (entry.second->GetWidthZ() * 0.5f));
+    }
+
+    areBoundariesSet = true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool MLPBaseHierarchyTool::IsInFV(const CartesianVector &position) const
+{
+    if ((position.GetX() < m_detectorMinX) or (position.GetX() > m_detectorMaxX))
+        return false;
+
+    if ((position.GetY() < m_detectorMinY) or (position.GetY() > m_detectorMaxY))
+        return false;
+
+    if ((position.GetZ() < m_detectorMinZ) or (position.GetZ() > m_detectorMaxZ))
+        return false;
```

Comment:
Not a review comment...but huh I'd never come across that you could use `or` / `and` rather than `||`/`&&` in C++!

Reply:
So ashamed.. I’ve been switching between python and C++ a lot recently 😅 I’ll change it to ||